### PR TITLE
Feature/coloring road

### PR DESCRIPTION
**Pull request by p0l0satik**
added road labels from kitti

### CODE REVIEW

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: loader/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import numpy as np
+from torch.utils.data import Dataset
+from loader.laserscan import SemLaserScan
+
+class KittiSimple(Dataset):
+    def __init__(self, dir, len) -> None:
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/dataloader.py]: fixed

File: loader/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        super().__init__()
+        self.ls = SemLaserScan()
+        self.len = len
+        self.label_folder = dir + "plane_labels/"
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/dataloader.py]: fixed

File: loader/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        self.road_label_folder = dir + "labels/"
+        self.file_folder = dir + "velodyne/"
+    def __len__(self):
+        return self.len
+    
+    def __getitem__(self, idx):
+        KITI_ROAD_LABELS = (40, 48, 44, 49)
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/dataloader.py]: fixed

File: loader/laserscan.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
+import numpy as np
+from torch import int16
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py]: fixed

File: loader/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        self.ls.open_scan(self.file_folder + '{0:06d}.bin'.format(idx))
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/dataloader.py]: fixed

File: loader/laserscan.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+import open3d as o3d
+
+class LaserScan:
+  """Class that contains LaserS
+  can with x,y,z,r"""
+  EXTENSIONS_SCAN = ['.bin', ".pcd", ".npy", ".label"]
+
+  def __init__(self,  H=64, W=1024, fov_up=3.0, fov_down=-25.0):
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py]: fixed

File: loader/laserscan.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                              dtype=np.int32)       # [H,W] mask
+
+  def size(self):
+    """ Return the size of the point cloud. """
+    return self.points.shape[0]
+
+  def __len__(self):
+    return self.size()
+
+  def open_scan(self, filename):
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py]: fixed

File: loader/laserscan.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    # if all goes well, open pointcloud
+    if filename.endswith(self.EXTENSIONS_SCAN[0]):
+      scan = np.fromfile(filename, dtype=np.float32)
+      scan = scan.reshape((-1, 4))
+      points = scan[:, 0:3]    # get xyz
+
+    elif filename.endswith(self.EXTENSIONS_SCAN[1]):
+      pcd = o3d.io.read_point_cloud(filename)
+      points = np.asarray(pcd.points)    # get xyz
+    else:
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py:66]: fixed

File: loader/laserscan.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    proj_y *= self.proj_H                              # in [0.0, H]
+
+    # round and clamp for use as index
+    proj_x = np.floor(proj_x)
+    proj_x = np.minimum(self.proj_W - 1, proj_x)
+    proj_x = np.maximum(0, proj_x).astype(np.int32)   # in [0,W-1]
+    self.proj_x = np.copy(proj_x)  # store a copy in orig order
+
+    proj_y = np.floor(proj_y)
+    proj_y = np.minimum(self.proj_H - 1, proj_y)
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py]: fixed

File: loader/laserscan.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+  def set_label(self, label):
+    """ Set points for label not from file but from np
+    """
+    # check label makes sense
+    if not isinstance(label, np.ndarray):
+      raise TypeError("Label should be numpy array")
+
+    # only fill in attribute if the right size
+    if label.shape[0] == self.points.shape[0]:
+      sem_label = label & 0xFFFF  # semantic label in lower half
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply loader/laserscan.py:161]: Arguably, but I insist on leaving bit masks and shifts as is because they are commented

File: visualisation.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        colors[:,0] = 255*labels
+        if show_projected_labels:
+            colors[proj_idx,1] = 255*proj_labels
+
+        pcd.points = o3d.utility.Vector3dVector(points)
+        pcd.colors = o3d.utility.Vector3dVector(colors)
+
+        o3d.visualization.draw_geometries([pcd])
+        break
+
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply visualisation.py:30]: fixed

File: visualisation.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import open3d as o3d
+import numpy as np
+from torch.utils.data import  DataLoader
+from loader.dataloader import *
+import time
+def visualise(loader, show_projected_labels = False):
```
[the remark this answers is not in the document]
Comment (p0l0satik) [reply visualisation.py]: fixed

**Review by kirill-ivanov-a — APPROVED** [review]
